### PR TITLE
T-5: Setup RTK & Setup json-server-auth

### PR DESCRIPTION
**Pull request by VladyslavGilyuk**
Registration & Login created and connected using Axios, json-server and json-auth
RTK & userSlicer installed and used to manage the state

UPD(25.09): 
Common Error - 'smth went wrong'
AsyncThunk => в action
extraReducers

### CODE REVIEW

File: db.json
Comment (mazhula97) [inline db.json:1]: It'll be better to keep this file empty.

File: src/redux/store.tsx
Comment (mazhula97) [inline src/redux/store.tsx:1]: rename redux folder to store. store.tsx -> index.ts

Comment (mazhula97) [reply src/redux/store.tsx:1]: keep structue like: store : { user, ... }: {slice.ts, actions.ts, selectors.ts}

File: src/hooks/useAuthForm.tsx [hunk 1]
```diff
@@ -18,17 +24,71 @@ export interface CommonFieldValues extends FieldValues {
   passwordConfirmation: string;
 }
 const useAuthForm = () => {
+  const dispatch = useDispatch();
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: need to add ts to the RTK, please read the doc

File: src/hooks/useAuthForm.tsx [hunk 2]
```diff
@@ -18,17 +24,71 @@ export interface CommonFieldValues extends FieldValues {
...
+  const navigate = useNavigate();
   const {
     handleSubmit,
     register,
     formState: { errors },
   } = useForm<CommonFieldValues>();
 
   const onSignInSubmit: SubmitHandler<ISignIn> = (data: ISignIn) => {
-    console.log('Form data:', data);
+    axios
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: Should be in asyncThunk, please read the doc

File: src/hooks/useAuthForm.tsx [hunk 3]
```diff
@@ -18,17 +24,71 @@ export interface CommonFieldValues extends FieldValues {
...
+      .post('/login', data)
+      .then(({ data }) => {
+        dispatch(
+          setUser({
+            token: data.accessToken,
+            ...data.user,
+          }),
+        );
+        localStorage.setItem(
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: As a option read about redux-persist, I do not insist but this library can help you to store your data in LS

File: src/redux/userSlice.tsx [hunk 4]
```diff
@@ -0,0 +1,27 @@
+import { createSlice } from '@reduxjs/toolkit';
+interface UserState {
```
Comment (mazhula97) [inline src/redux/userSlice.tsx]: It's time to create src/types folder with files like: user.ts, tickets.ts, etc. where we'll store all common types

File: src/hooks/useAuthForm.tsx [hunk 5]
```diff
@@ -18,17 +24,71 @@ export interface CommonFieldValues extends FieldValues {
   passwordConfirmation: string;
 }
 const useAuthForm = () => {
+  const dispatch = useDispatch();
+
+  useEffect(() => {
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: In general have no idea what is going on there.
This login check should works at the app initialization

File: src/utils/axios.ts
Comment (mazhula97) [inline src/utils/axios.ts:1]: We decided to create src/service folder with common axios server. For exmple: 
class UserService {
login(body) {
return axios.post('url', body);
}
}


then: asyncthunk({email, password}) -> UserService.login({email, password});

then: appDispatch(asyncThunk({email, password}))

**Review by mazhula97 — CHANGES_REQUESTED** [review]

File: src/redux/store.tsx
Comment (VladyslavGilyuk) [reply src/redux/store.tsx:1]: folder renamed, files rearranged

File: src/hooks/useAuthForm.tsx (hunk 1 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: TS generally added to RTK

File: src/redux/userSlice.tsx (hunk 4 above)
Comment (VladyslavGilyuk) [reply src/redux/userSlice.tsx]: types created, interfaces for users moved to user.ts

File: src/hooks/useAuthForm.tsx (hunk 3 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: Not changed at the moment but noted down

File: src/hooks/useAuthForm.tsx (hunk 2 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: asyncThunk added
Upd: specific errors (Invail email/password) not working

File: src/hooks/useAuthForm.tsx (hunk 5 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: useEffect deleted

File: db.json
Comment (VladyslavGilyuk) [reply db.json:1]: db emptied

File: src/utils/axios.ts
Comment (VladyslavGilyuk) [reply src/utils/axios.ts:1]: UserService with login and register created and later used

File: src/hooks/useAuthForm.tsx [hunk 6]
```diff
@@ -1,35 +1,39 @@
...
     handleSubmit,
     register,
     formState: { errors },
-  } = useForm<CommonFieldValues>();
+  } = useForm<ICommonFieldValues>();
 
-  const onSignInSubmit: SubmitHandler<ISignIn> = (data: ISignIn) => {
-    console.log('Form data:', data);
+  const onSignInSubmit: SubmitHandler<ISignIn> = async (data: ISignIn) => {
+    await dispatch(signInAsync(data))
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: is `await` necessary in this place?

File: src/hooks/useAuthForm.tsx [hunk 7]
```diff
@@ -1,35 +1,39 @@
...
+        console.log(error);
+      });
   };
 
-  const onSignUpSubmit: SubmitHandler<ISignUp> = (data: ISignUp) => {
-    console.log('Form data:', data);
+  const onSignUpSubmit: SubmitHandler<ISignUp> = async (data: ISignUp) => {
+    const { email, password, firstName, lastName } = data;
+    const body = { email, password, firstName, lastName };
+    await dispatch(signUpAsync(body))
```
Comment (mazhula97) [inline src/hooks/useAuthForm.tsx]: is `await` necessary in this place?

File: src/store/user/actions.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+import { ISignIn, ISignUpAsync } from '../../types/auth';
+import { createAction, createAsyncThunk } from '@reduxjs/toolkit';
+
+export const setUser = createAction<UserState>('user/setUser');
+
+export const signInAsync = createAsyncThunk('auth/signIn', async (data: ISignIn) => {
+  try {
+    const response = await UserService.login(data);
+    const responseData = response.data;
+    console.log(responseData);
```
Comment (mazhula97) [inline src/store/user/actions.ts]: console.log. Please add eslint rule for console logs

File: src/store/user/actions.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    console.log(error);
+    alert('An error occurred. Please try again later.');
+  }
+});
+
+export const signUpAsync = createAsyncThunk('auth/signUp', async (data: ISignUpAsync) => {
+  try {
+    const response = await UserService.register(data);
+    const responseData = response.data;
+    console.log(responseData);
```
Comment (mazhula97) [inline src/store/user/actions.ts]: console.log

File: src/store/user/slice.tsx [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+export const userSlice = createSlice({
+  name: 'user',
+  initialState,
+  reducers: {},
+  extraReducers(builder) {
+    builder.addCase(signInAsync.fulfilled, (state, action) => {
+      const responseData = action.payload;
+      return {
+        token: responseData.accessToken,
+        ...responseData.user,
```
Comment (mazhula97) [inline src/store/user/slice.tsx]: Will we have password field in our redux store?

**Review by mazhula97 — CHANGES_REQUESTED** [review]

File: src/hooks/useAuthForm.tsx (hunk 6 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: await changed to .then()

File: src/hooks/useAuthForm.tsx (hunk 7 above)
Comment (VladyslavGilyuk) [reply src/hooks/useAuthForm.tsx]: await changed to .then()

File: src/store/user/slice.tsx (hunk 10 above)
Comment (VladyslavGilyuk) [reply src/store/user/slice.tsx]: No we won't.

File: src/store/user/actions.ts (hunk 8 above)
Comment (VladyslavGilyuk) [reply src/store/user/actions.ts]: "no-console": "error" added to eslint

File: src/store/user/actions.ts (hunk 9 above)
Comment (VladyslavGilyuk) [reply src/store/user/actions.ts]: all console.logs removed

**Review by mazhula97 — APPROVED** [review]

**Pull request merged**